python/tank/platform/__init__.py

```diff
@@ -77,7 +77,42 @@ def _get_current_bundle():
 ################################################################################################
 # Public API methods
 
-def restart():
+def change_context(new_context):
+    """
+    Running change_context will attempt to change the context the engine and
+    its apps are running in on the fly. The current engine must accept the
+    context change, otherwise a full restart of the engine will be run instead.
+
+    Any apps that support context switching on the fly will do so. Any that do
+    not will be restarted.
+
+    The benefit of supporting context changes in engines and apps is speed. The
+    end result of this routine should be identical to that of a restart, but
+    will require less time to complete.
+
+    For more information on supporting context changing, see the following:
+
+        tank.platform.engine.context_change_allowed
+        tank.platform.application.context_change_allowed
+        tank.platform.engine.change_context()
+        tank.platform.application.change_context()
+    """
+    engine = current_engine()
+
+    if engine is None:
+        raise TankError("No engine is currently running! Run start_engine instead.")
+
+    if not isinstance(new_context, engine.context.__class__):
```

Comment:
Generally speaking, curious if we need this at all - this concept of type checking parameters is nothing we do in toolkit, in other places we just assume that the right object is passed and rely on the code to raise further down. I guess it's kind of a case of "ask for forgiveness, rather than permission".

In some places, as part of the "public" toolkit API, we do validation and sometimes we do it just to be able to provide a better error message. But curious about this case - do you think we should generally be stricter with this? Check all parameters for all public API methods? That feels a little excessive to me perhaps. Alternatively, I think we might want to add a comment explaining why we are validating in this case but not in others.

I am guessing if we skip validation, an error will bubble up from engine.change_context() - but maybe this error msg is confusing and/or ugly looking?